File: src/main/java/me/testaccount666/serversystem/clickablesigns/executables/kit/ActionKitSign.java
package me.testaccount666.serversystem.clickablesigns.executables.kit;

import me.testaccount666.serversystem.ServerSystem;
import me.testaccount666.serversystem.clickablesigns.AbstractSignClickAction;
import me.testaccount666.serversystem.commands.executables.kit.manager.KitManager;
import me.testaccount666.serversystem.userdata.User;
import me.testaccount666.serversystem.utils.ChatColor;
import org.bukkit.block.Sign;
import org.bukkit.configuration.file.FileConfiguration;

import static me.testaccount666.serversystem.utils.MessageBuilder.sign;

public class ActionKitSign extends AbstractSignClickAction {

    @Override
    public String getBasePermissionNode() {
        return "ClickableSigns.Kit";
    }

    @Override
    protected boolean executeAction(User user, Sign sign, FileConfiguration config) {
        var kitManager = ServerSystem.Instance.getRegistry().getService(KitManager.class);
        if (kitManager == null) {
            sign("Kit.NoKitManager", user).build();
            return false;
        }

        var kitName = config.getString("KitName", sign.getLine(1));
        kitName = ChatColor.stripColor(kitName);
        if (kitName.isEmpty()) {
            sign("Kit.NoKitSpecified", user).build();
            return false;
        }

        var kitOptional = kitManager.getKit(kitName.toLowerCase());
        if (kitOptional.isEmpty()) {
            var finalKitName = kitName;
            sign("Kit.KitNotFound", user)
                    .postModifier(message -> message.replace("<KIT>", finalKitName)).build();
            return false;
        }

        var kit = kitOptional.get();
        kit.giveKit(user.getPlayer());
        sign("Kit.KitGiven", user)
                .postModifier(message -> message.replace("<KIT>", kit.getDisplayName())).build();
        return true;
    }
}

File: src/main/kotlin/me/testaccount666/serversystem/clickablesigns/executables/kit/ActionKitSign.kt
package me.testaccount666.serversystem.clickablesigns.executables.kit

import me.testaccount666.paperktx.colors.ChatColor.Companion.stripColor
import me.testaccount666.serversystem.clickablesigns.AbstractSignClickAction
import me.testaccount666.serversystem.commands.executables.kit.manager.KitManager
import me.testaccount666.serversystem.extensions.*
import me.testaccount666.serversystem.userdata.User
import me.testaccount666.serversystem.utils.DurationParser.parseDate
import org.bukkit.block.Sign
import org.bukkit.configuration.file.FileConfiguration

class ActionKitSign : AbstractSignClickAction() {
    override val basePermissionNode = "ClickableSigns.Kit"

    override fun executeAction(user: User, sign: Sign, config: FileConfiguration, onSuccess: () -> Unit): Boolean {
        val kitManager = getService<KitManager>()

        val kitName = stripColor(config.getString("KitName", sign.getLine(1))).takeUnless { it.isEmpty() } ?: run {
            user.signMsg("Kit.NoKitSpecified")
            return false
        }

        val kit = kitManager.getKit(kitName) ?: run {
            user.signMsg("Kit.KitNotFound") {
                postModifier { it.replace("<KIT>", kitName) }
            }
            return false
        }

        if (user.isOnKitCooldown(kit.name)) {
            val cooldown = user.getKitCooldown(kitName)

            user.commandMsg("Kit.OnCooldown") {
                postModifier {
                    it.replace("<KIT>", kit.displayName)
                        .replace("<DATE>", parseDate(cooldown, user))
                }
            }
            return false
        }

        user.setKitCooldown(kit.name, kit.coolDown).also { user.save() }

        kit.giveKit(user.getPlayer()!!)
        user.signMsg("Kit.KitGiven") {
            postModifier { it.replace("<KIT>", kit.displayName) }
        }
        onSuccess()
        return true
    }
}